apps/tailwind-components/tests/vitest/composables/useForm.spec.ts

```diff
@@ -0,0 +1,136 @@
+import { describe, expect, test, vi } from "vitest";
+import type { ITableMetaData } from "../../../../metadata-utils/src";
+import type {
+  columnId,
+  columnValue,
+} from "../../../../metadata-utils/src/types";
+
+describe("useForm", () => {
+  const tableMetadata: ITableMetaData = {
+    id: "vi test table metadata",
+    label: "vi test table metadata",
+    tableType: "some table type",
+    columns: [
+      {
+        columnType: "STRING",
+        id: "col1",
+        label: "columns 1",
+      },
+      {
+        columnType: "STRING",
+        id: "col2",
+        label: "columns 2",
+        required: true,
+      },
+      {
+        columnType: "STRING",
+        id: "col3",
+        label: "columns 3",
+      },
+      {
+        columnType: "STRING",
+        id: "col4",
+        label: "columns 4",
+        required: true,
+      },
+    ],
+  };
+
+  const formValues = ref<Record<string, columnValue>>({});
+  const errorMap = ref<Record<columnId, string>>({});
+  const scrollTo = vi.fn();
+
+  const {
+    requiredFields,
+    emptyRequiredFields,
+    requiredMessage,
+    errorMessage,
+    gotoPreviousRequiredField,
+    gotoNextRequiredField,
+    gotoNextError,
+    gotoPreviousError,
+  } = useForm(tableMetadata, formValues, errorMap, scrollTo);
+
+  test("should return a list of required fields", () => {
+    expect(requiredFields.value).toEqual([
+      {
+        columnType: "STRING",
+        id: "col2",
+        label: "columns 2",
+        required: true,
+      },
+      {
+        columnType: "STRING",
+        id: "col4",
+        label: "columns 4",
+        required: true,
+      },
+    ]);
+  });
+
+  test("should return a list of empty required fields", () => {
+    expect(emptyRequiredFields.value).toEqual([
+      {
+        columnType: "STRING",
+        id: "col2",
+        label: "columns 2",
+        required: true,
+      },
+      {
+        columnType: "STRING",
+        id: "col4",
+        label: "columns 4",
+        required: true,
+      },
+    ]);
+  });
+
+  test("should go to the next required field", () => {
+    gotoNextRequiredField();
+    expect(scrollTo).toHaveBeenCalledWith("col2-form-field");
+    gotoNextRequiredField();
+    expect(scrollTo).toHaveBeenCalledWith("col4-form-field");
+  });
+
+  test("should go to the previous required field", () => {
+    gotoPreviousRequiredField();
+    expect(scrollTo).toHaveBeenCalledWith("col2-form-field");
+    gotoPreviousRequiredField();
+    expect(scrollTo).toHaveBeenCalledWith("col4-form-field");
+  });
+
```

Comment:
test are not independent, I can comment line 96 and 98 and the test still succeeds because of the test before it.